README.md
# Baby Bitbox

*NOTE* - this is a fork of what appears to be an abandoned library updated with current deps.   Updated versions are tagged as baby*.  

Lite version of [Bitcoin.com's Bitbox JS library](https://developer.bitcoin.com/bitbox/) for building Bitcoin Cash apps in Flutter. Integrated to Bitcoin.com's REST API.

Want the original?  It's [here](https://pub.dev/packages/bitbox)

Works with mainnet and testnet.  Pointycastle does not support Web - but this should run on other platforms.

## Getting Started

### 1) Depend on it
If you just want to get this from Dart's public package directory:

```
dependencies:
  babybitbox: ^0.1.1
```

If you checked this out from Github, add a local dependency into the pubspec.yaml of your testing or development projet:

```
dependencies:
  babybitbox_plugin:
    path: <path to the directory>/
```
	
### 2) Import it

```
// There's a good chance your own project will use similar names as some of the 
// classes in this library. A simple way to create some order is to import the 
// library with Bitbox prefix:
import 'package:babybitbox/babybitbox.dart' as Bitbox;
```

### 2) Use it
```
// set this to true to use testnet
final testnet = true;

// After running the code for the first time, depositing an amount to the address 
// displayed in the console, and waiting for confirmation, paste the generated 
// mnemonic here, so the code continues below with address withdrawal
String mnemonic = "";

if (mnemonic == "") {
  // generate 12-word (128bit) mnemonic
  mnemonic = Bitbox.Mnemonic.generate();

  print(mnemonic);
}

// generate a seed from mnemonic
final seed = Bitbox.Mnemonic.toSeed(mnemonic);

// create an instance of Bitbox.HDNode for mainnet
final masterNode = Bitbox.HDNode.fromSeed(seed, testnet);

// This format is compatible with Bitcoin.com wallet.
// Other wallets use Change to m/44'/145'/0'/0
final accountDerivationPath = "m/44'/0'/0'/0";

// create an account node using the provided derivation path
final accountNode = masterNode.derivePath(accountDerivationPath);

// get account's extended private key
final accountXPriv = accountNode.toXPriv();

// create a Bitbox.HDNode instance of the first child in this account
final childNode = accountNode.derive(0);

// get an address of the child
final address = childNode.toCashAddress();

// if you are using testnet, set the appropriate rest api url before making any API
// calls (like getting address or transaction details or broadcasting a transaction
if (testnet) {
  Bitbox.Bitbox.setRestUrl(restUrl: Bitbox.Bitbox.trestUrl);
}

// get address details
final addressDetails = await Bitbox.Address.details(address);

print(addressDetails);

// If there is a confirmed balance, attempt to withdraw it
if (addressDetails["balance"] > 0) {
  final builder = Bitbox.Bitbox.transactionBuilder(testnet: testnet);

  // retrieve address' utxos from the rest api
  final utxos = await Bitbox.Address.utxo(address) as List<Bitbox.Utxo>;

  // placeholder for input signatures
  final signatures = <Map>[];

  // placeholder for total input balance
  int totalBalance = 0;

  // iterate through the list of address utxos and use them as inputs for the 
  // withdrawal transaction
  utxos.forEach((Bitbox.Utxo utxo) {
    // add the utxo as an input for the transaction
    builder.addInput(utxo.txid, utxo.vout);

    // add a signature to the list to be used later
    signatures.add({
      "vin": signatures.length,
      "key_pair": childNode.keyPair,
      "original_amount": utxo.satoshis
    });

    totalBalance += utxo.satoshis;
  });

  // set an address to send the remaining balance to
  final outputAddress = "";

  // if there is an unspent balance, create a spending transaction
  if (totalBalance > 0 && outputAddress != "") {
    // calculate the fee based on number of inputs and one expected output
    final fee = Bitbox.BitcoinCash.getByteCount(signatures.length, 1);

    // calculate how much balance will be left over to spend after the fee
    final sendAmount = totalBalance - fee;

    // add the output based on the address provided in the testing data
    builder.addOutput(outputAddress, sendAmount);

    // sign all inputs
    signatures.forEach((signature) {
      builder.sign(signature["vin"], signature["key_pair"], 
	    signature["original_amount"]);
    });

    // build the transaction
    final tx = builder.build();

    // broadcast the transaction
    final txid = await Bitbox.RawTransactions.sendRawTransaction(tx.toHex());

    // Yatta!
    print("Transaction broadcasted: $txid");
  } else if (totalBalance > 0) {
    print("Enter an output address to test withdrawal transaction");
  }
}
```

Memo.cash is now supported - but with a limited subset (many of the calls aren't easily supportable - if you need can add an issue).

```
    BabyBitbox.Memo.setName(childNode, toThis);
```

For further documentation, refer to apidoc of this repository

## Testing

There are some unit tests in test/bitbox_test.dart. They use data generated from the original [Bitbox for JS](https://developer.bitcoin.com/bitbox/) and compare them with the output of this library. 
The following is tested for both testnet and mainnet:
- Generating the master node from mnemonic and comparing both its XPub and XPriv
- Generating an account node and comparing XPub and XPriv
- Generating 10 test childs and comparing their private keys and addresses
- Conversion of cashAddr format to legacy and vice versa
- Fetching address details for the rest API
- Fetching utxos for addresses with balance
- Building a transaction spending the deposited balance and comparing its hash with Bitbox JS output
- Optionally broadcasting the transaction and

To run the test:

1. Copy create_test_data.js to a separate directory and download the original Bitbox JS into the directory
2. Generate the testing data by runing create_test_data.js with your local nodeJS engine
3. Update bitbox_test.dart with the path to the generated test_data.json file
4. Run bitbox_test.dart 
Optionally between step 1) and 2), send some balance to either testnet or mainnet addresses (or both), wait for confirmations and run create_test_data.js again to update the data and generate testing transactions


## Acknowledgments

This version is a forked version of https://pub.dev/packages/bitbox - Tomas did all the heavy lifting and deserves all the credit for goodness in this library, I own all the bugs.


CHANGELOG.md
## [0.0.1] - 2019-08-01

Light version of Bitcoin.com's Bitbox library

## [0.0.2] - 2019-10-06

- Taken some code analysis hint into account

## [0.0.3] - 2020-03-20

- Added support for bip21 implemented by RomitRadical
- Added transaction details support
- Fixed some API issues and

## [0.0.4] - 2020-05-12

- padded a private key by if it was generated shorter than 32 bytes
- added support for broadcasting more transactions
- changed resturl parameter to non-named (didn't make sense to have it named when it was the only one))

## [0.0.4] - 2020-06-02

- fixed signing mechanism, so it would better handle 31-byte private key

## [0.1.0] - 2020-11-20
- update deps, pointycastle to current versions

## [0.1.1] - 2020-11-29
- memo.cash updates



test/bitbox_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

import 'package:babybitbox/babybitbox.dart' as Bitbox;

// These tests generate a bunch of keys, addresses and compare them to the testing data generated by the original bitbox
// If there are balances on the addresses, the tests retrieve utxos, compile testing spending transactions,
// and optionally broadcast the transactions
//
// Make sure to run create_test_data.js first to generate the test data.
void main() {
  // path to the file generated by create_test_data.js
  const TEST_DATA_PATH = "/tmp/test_data.json";

  // If these are false, the transactions will be only built and compared to the output generated by bitbox js
  // You can turn these on separately
  const BROADCAST_TESTNET_TRANSACTION = false;
  const BROADCAST_MAINNET_TRANSACTION = false;

  // Data generated by the original bitbox library
  Map testData;

  // Placeholder for data about master, account and childnodes for both networks
  Map nodeData = {"mainnet" : {}, "testnet" : {}};

  // The whole code would be a bit more elegant if this was a map with true/false as keys and forEach would be used
  // in each of the testing methods. However the whole asynchronisity is messed up when it's done that way,
  // so a simple list and for() loop is used
  final networks = ["mainnet", "testnet"];

  // Generate master node master private key and public key and compare with the testing data
  test('Generating master node from mnemonic', () {
    // retrieve and parse the testing data
    final testFile = File.fromUri(Uri(path: TEST_DATA_PATH));
    final testDataJson = testFile.readAsStringSync();
    testData = jsonDecode(testDataJson);

    // create a seed from the mnemonic
    final seed = Bitbox.Mnemonic.toSeed(testData["mnemonic"]);

    // create master nodes for both networks and store their master keys
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];
      nodeData[network]["master_node"] = Bitbox.HDNode.fromSeed(seed, network == "testnet");
      final masterXPriv = nodeData[network]["master_node"].toXPriv();
      final masterXpub = nodeData[network]["master_node"].toXPub();

      // compare the result with the js test data
      expect(masterXPriv, testData[network]["master_xpriv"]);
      expect(masterXpub, testData[network]["master_xpub"]);
    }
  });

  // generate account node and compare its master keys with the original testing data
  test('Generating account node', () {
    // generate the nodes for both networks
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];
      nodeData[network]["account_node"] = nodeData[network]["master_node"].derivePath(testData["account_path"]);
      final accountXPriv = nodeData[network]["account_node"].toXPriv();
      final accountXPub = nodeData[network]["account_node"].toXPub();

      // compare the master private and public key with the original testing data
      expect(accountXPriv, testData[network]["account_xpriv"]);
      expect(accountXPub, testData[network]["account_xpub"]);
    }
  });

  // The following few methods work with child nodes of the account node created above.
  // Tests to generate various outputs from each child nodes are separated
  // It is not determined here how many child nodes are created.
  // The test simply follows the index and thus number of child nodes stored in the original test file.
  // That being said, it takes only the index number and derives all other data itself.

  // First generate private keys for each of the child nodes
  test('Generating child nodes and private keys', () {
    // Generate child nodes for each network
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];

      testData[network]["child_nodes"].forEach((childTestData) {
        // generate the child node and extract its private key
        final childNode = nodeData[network]["account_node"].derive(childTestData["index"]);
        final childPrivateKey = childNode.privateKey;

        // compare the private key with the original test file
        expect(childPrivateKey, childTestData["private_key"]);
      });
    }
  });

  test('Generating child nodes and legacy addresses', () {
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];
      testData[network]["child_nodes"].forEach((childTestData) {
        final childNode = nodeData[network]["account_node"].derive(childTestData["index"]);
        final childLegacy = childNode.toLegacyAddress();

        expect(childLegacy, childTestData["legacy"]);
      });
    }
  });

  test('Generating child nodes and cash addresses', () {
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];
      testData[network]["child_nodes"].forEach((childTestData) {
        final childNode = nodeData[network]["account_node"].derive(childTestData["index"]);
        final childCashAddr = childNode.toCashAddress();

        expect(childCashAddr, childTestData["cashAddress"]);
      });
    }
  });

  // For these conversion tests the script uses the addresses from the original script
  test('Converting cashAddr to legacy', () {
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];
      testData[network]["child_nodes"].forEach((childTestData) {
        final cashAddr = childTestData["cashAddress"];

        expect(Bitbox.Address.toLegacyAddress(cashAddr), childTestData["toLegacy"]);
      });
    }
  });

  test('Converting legacy to cashAddr', () {
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];
      testData[network]["child_nodes"].forEach((childTestData) {
        final legacy = childTestData["legacy"];

        expect(Bitbox.Address.toCashAddress(legacy), childTestData["toCashAddr"]);
      });
    }
  });

  // Placeholder to store addresses with balance for which to fetch utxos later
  final utxosToFetch = <String, List<String>>{};

  test('Fetching address details', () async {
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];
      utxosToFetch[network] = <String>[];

      // set rest url based on which network is being tested
      Bitbox.Bitbox.setRestUrl(network == "mainnet" ? Bitbox.Bitbox.restUrl : Bitbox.Bitbox.trestUrl);

      // Placeholder for test addresses to fetch the details off
      List<String> testAddresses = <String>[];

      // Accumulate the list of all addresses from the test file
      testData[network]["child_nodes"].forEach((childTestData) {
        testAddresses.add(childTestData["cashAddress"]);
      });

      // test retreiving both single address details and list of all addresses
      final detailsSingle = await Bitbox.Address.details(testAddresses.first);
      final detailsAll = await Bitbox.Address.details(testAddresses) as List;

      // check if the return data is of the expected type
      expect(true, detailsSingle is Map);
      expect(true, detailsAll is List);

      // store all addresses with non-zero confirmed balance
      detailsAll.forEach((addressDetails) {
        if (addressDetails["balance"] > 0) {
          utxosToFetch[network].add(addressDetails["cashAddress"]);
        }
      });
    }
  });

  // Placeholder for utxo details for later use
  Map<String, List> utxos = {};

  test('Fetching utxos', () async {
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];

      // placeholder for utxos for the particular network
      utxos[network] = [];

      // If there were addresses with non-zero balance for this network, fetch their utxos
      if (utxosToFetch[network].length > 0) {
        // set the appropriate rest api url
        Bitbox.Bitbox.setRestUrl(network == "mainnet" ? Bitbox.Bitbox.restUrl : Bitbox.Bitbox.trestUrl);

        // fetch the required utxo details
        utxos[network] = await Bitbox.Address.utxo(utxosToFetch[network]) as List;

        // go through the list of the returned utxos
        utxos[network].forEach((addressUtxo) {
          // go through each address in the testing data to test if the utxos match
          testData[network]["child_nodes"].forEach((childNode) {
            if (childNode["cashAddress"] == addressUtxo["cashAddress"]) {
              for (int i = 0; i < addressUtxo["utxos"].length; i++) {
                Bitbox.Utxo utxo = addressUtxo["utxos"][i];

                expect(utxo.txid, childNode["utxos"][i]["txid"]);
              }
            }
          });
        });
      }
    }
  });

  // Placeholder for both mainnet and testnet raw transaction data
  Map<String, String> rawTx = {};

  // If there are unspent outputs for any of the addresses in any of the network, built spending transactions and
  // compare them with the testing data
  test('Building spending transaction', () async {
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];

      // Placeholder for total unspent balance
      int totalBalance = 0;

      // placeholder for input signatures
      final signatures = <Map>[];

      // create a transaction builder for the appropriate network
      final builder = Bitbox.Bitbox.transactionBuilder(testnet: network == "testnet");

      // go through the list of utxos accumulated in the previous test
      utxos[network].forEach((addressUtxos) {
        testData[network]["child_nodes"].forEach((childNode) {
          if (childNode["cashAddress"] == addressUtxos["cashAddress"]) {
            addressUtxos["utxos"].forEach((Bitbox.Utxo utxo) {
              // add the utxo as an input for the transaction
              builder.addInput(utxo.txid, utxo.vout);

              // add a signature to the list to be used later
              signatures.add({
                "vin": signatures.length,
                "key_pair": nodeData[network]["account_node"].derive(childNode["index"]).keyPair,
                "original_amount": utxo.satoshis
              });

              totalBalance += utxo.satoshis;
            });
          }
        });
      });

      // if there is an unspent balance, create a spending transaction
      if (totalBalance > 0) {
        // calculate the fee based on number of inputs and one expected output
        final fee = Bitbox.BitcoinCash.getByteCount(signatures.length, 1);

        // calculate how much balance will be left over to spent after the fee
        final sendAmount = totalBalance - fee;

        // add the ouput based on the address provided in the testing data
        builder.addOutput(testData[network]["output_address"], sendAmount);

        // sign all inputs
        signatures.forEach((signature) {
          builder.sign(signature["vin"], signature["key_pair"], signature["original_amount"]);
        });

        // build the transaction
        final tx = builder.build();

        // compare the transaction raw hex with the output from the original bitbox
        expect(tx.toHex(), testData[network]["testing_tx_hex"]);

        // add the raw transaction to the list to be (optionally) broadcastd
        rawTx[network] = tx.toHex();
      }
    }
  });

  // broadcast the transactions if the appropriate flags have been raised
  test('Broadcasting the transaction', () async {
    for (int i = 0; i < networks.length; i++) {
      final network = networks[i];

      // check if there is a transaction for this network
      if (rawTx.containsKey(network)) {
        // check if this transaction is supposed to be broadcasted
        if ((network == "testnet" && BROADCAST_TESTNET_TRANSACTION)
            || (network == "mainnet" && BROADCAST_MAINNET_TRANSACTION)) {
          // set the appropraite rest api url
          Bitbox.Bitbox.setRestUrl(network == "mainnet" ? Bitbox.Bitbox.restUrl : Bitbox.Bitbox.trestUrl);
          // broadcast the transaction and print its id
          final txid = (await Bitbox.RawTransactions.sendRawTransaction([rawTx[network]])).first;

          expect(true, txid is String);
        }
      }
    }
  });

  test('BIP21 test', () {
    final address = 'bitcoincash:qrdsfshx7yzfjl9sfj2khuja5crcu4vaxqrt2qkz5s';

    final options = {
      "amount": 1.0,
      "label": '#BCHForEveryone',
      "message": "There you go"
    };

    final encodedPaymentRequest = "bitcoincash:qrdsfshx7yzfjl9sfj2khuja5crcu4vaxqrt2qkz5s?"
      +"amount=1.0&label=%23BCHForEveryone&message=There%20you%20go";

    expect(Bitbox.BitcoinCash.encodeBIP21(address, options), encodedPaymentRequest);

    final decodedPaymentRequest = Bitbox.BitcoinCash.decodeBIP21(encodedPaymentRequest);

    expect(decodedPaymentRequest["address"], address);
    expect(decodedPaymentRequest["options"]["amount"], options["amount"]);
    expect(decodedPaymentRequest["options"]["label"], options["label"]);
    expect(decodedPaymentRequest["options"]["message"], options["message"]);
  });
}


lib/src/ecpair.dart
import 'dart:typed_data';
import 'dart:math';
import 'package:babybip32/src/utils/ecurve.dart' as ecc;
import 'package:babybip32/src/utils/wif.dart' as wif;
import 'package:pointycastle/ecc/api.dart';

import 'address.dart';
import 'crypto/crypto.dart';
import 'utils/network.dart';

/// Stores a keypair and provides various methods and factories for creating it and working with it
class ECPair {
  final Uint8List _d;
  final Uint8List _q;
  final Network network;
  final bool compressed;

  /// Default constructor. If [network] is not provided, it will assume Bitcoin Cash mainnet
  ECPair(this._d, this._q, {network, this.compressed = true}):
    this.network = network ?? Network.bitcoinCash();

  /// Creates a keypair from the private key provided in WIF format
  factory ECPair.fromWIF(String wifPrivateKey, {Network network}) {
    wif.WIF decoded = wif.decode(wifPrivateKey);
    final version = decoded.version;
    // TODO support multi networks
    Network nw;
    if (network != null) {
      nw = network;
      if (nw.private != version) throw ArgumentError("Invalid network version");
    } else {
      if (version == Network.bchPrivate) {
        nw = Network.bitcoinCash();
      } else if (version == Network.bchTestnetPrivate) {
        nw = Network.bitcoinCashTest();
      } else {
        throw ArgumentError("Unknown network version");
      }
    }
    return ECPair.fromPrivateKey(decoded.privateKey, compressed: decoded.compressed, network: nw);
  }

  /// Creates a keypair from [publicKey. The returned keypair will contain [null] private key
  factory ECPair.fromPublicKey(Uint8List publicKey, {Network network, bool compressed}) {
    if (!ecc.isPoint(publicKey)) {
      throw ArgumentError("Point is not on the curve");
    }
    return ECPair(null, publicKey,
      network: network, compressed: compressed);
  }

  /// Creates a keypair from [privateKey]
  factory ECPair.fromPrivateKey(Uint8List privateKey, {Network network, bool compressed}) {
    if (privateKey.length != 32)
      throw ArgumentError(
        "Expected property privateKey of type Buffer(Length: 32)");
    if (!ecc.isPrivate(privateKey))
      throw ArgumentError("Private key not in range [1, n)");
    return ECPair(privateKey, null,
      network: network, compressed: compressed);
  }

  /// Creates a random keypair
  factory ECPair.makeRandom({Network network, bool compressed, Function rng}) {
    final rfunc = rng ?? _randomBytes;
    Uint8List d;
//    int beginTime = DateTime.now().millisecondsSinceEpoch;
    do {
      d = rfunc(32);
      if (d.length != 32) throw ArgumentError("Expected Buffer(Length: 32)");
//      if (DateTime.now().millisecondsSinceEpoch - beginTime > 5000) throw ArgumentError("Timeout");
    } while (!ecc.isPrivate(d));
    return ECPair.fromPrivateKey(d, network: network, compressed: compressed);
  }

  Uint8List get publicKey  => _q ?? ecc.pointFromScalar(_d, compressed);

  Uint8List get privateKey => _d;

  String get address => Address.toBase58Check(hash160(publicKey), network.pubKeyHash);

  /// Returns the private key in WIF format
  String toWIF() {
    if (privateKey == null) {
      throw ArgumentError("Missing private key");
    }
    return wif.encode(wif.WIF(
      version: network.private, privateKey: privateKey, compressed: compressed));
  }

  /// Returns signature object instead of serialized signature so that it is better handled in bip66 encoding
  ECSignature sign(Uint8List hash) {
    return eccSign(hash, privateKey);
  }

  /// Verifies whether the provided [signature] matches the [hash] using the keypair's [publicKey]
  bool verify(Uint8List hash, Uint8List signature) {
    return ecc.verify(hash, publicKey, signature);
  }

  /// custom EC signature function that returns a signature object instead of serialized result
  ECSignature eccSign(Uint8List hash, Uint8List x) {
    if (!ecc.isScalar(hash)) throw new ArgumentError(ecc.THROW_BAD_HASH);
    if (!ecc.isPrivate(x)) throw new ArgumentError(ecc.THROW_BAD_PRIVATE);

    ECSignature sig = ecc.deterministicGenerateK(hash, x);

    return sig;
  }

}

const int _SIZE_BYTE = 255;
Uint8List _randomBytes(int size) {
  final rng = Random.secure();
  final bytes = Uint8List(size);
  for (var i = 0; i < size; i++) {
    bytes[i] = rng.nextInt(_SIZE_BYTE);
  }
  return bytes;
}


lib/src/utils/bip21.dart
import 'package:babybitbox/babybitbox.dart';

class Bip21 {

  static Map<String, dynamic> decode(String uri) {
    if (uri.indexOf('bitcoincash') != 0 || uri['bitcoincash'.length] != ":")
      throw ("Invalid BIP21 URI");

    int split = uri.indexOf("?");
    Map<String, String> uriOptions = Uri.parse(uri).queryParameters;

    Map<String, dynamic> options = Map.from({
      "message": uriOptions["message"],
      "label": uriOptions["label"],
    });

    String address = uri.substring(0, split == -1 ? null : split);

    if (uriOptions["amount"] != null) {
      if (uriOptions["amount"].indexOf(",") != -1)
        throw ("Invalid amount: commas are invalid");

      double amount = double.tryParse(uriOptions["amount"]);
      if (amount == null || amount.isNaN)
        throw ("Invalid amount: not a number");
      if (!amount.isFinite) throw ("Invalid amount: not finite");
      if (amount < 0) throw ("Invalid amount: not positive");
      options["amount"] = amount;
    }

    return {
      'address': address,
      'options': options,
    };
  }

  static String encode(String address, Map<String, dynamic> options) {
    var hasCashPrefix = address.startsWith('bitcoincash:');
    if (!hasCashPrefix) {
      address = 'bitcoincash:$address';
    }

    // Using toLegacyAddress method to test validity of the address format
    try {
      Address.toLegacyAddress(address);
    } catch (e) {
      throw FormatException("Invalid Address Format: $address");
    }

    String query = "";
    if (options != null && options.isNotEmpty) {
      if (options['amount'] != null) {
        if (!options['amount'].isFinite) throw ("Invalid amount: not finite");
        if (options['amount'] < 0) throw ("Invalid amount: not positive");
      }

      Map<String, dynamic> uriOptions = Map.from(options);
      uriOptions.removeWhere((key, value) => value == null);
      uriOptions.forEach((key, value) {
        uriOptions[key] = value.toString();
      });

      if (uriOptions.isEmpty) uriOptions = null;
      query = Uri(queryParameters: uriOptions).toString();
      // Dart isn't following RFC-3986...
      query = query.replaceAll(RegExp(r"\+"), "%20");
    }

    return "$address$query";
  }
}


lib/src/utils/p2pkh.dart
import 'dart:typed_data';
import '../crypto/crypto.dart';
import '../utils/opcodes.dart';
import 'package:meta/meta.dart';
import 'package:babybip32/src/utils/ecurve.dart' show isPoint;
import 'package:bs58check/bs58check.dart' as bs58check;
import 'script.dart' as bscript;

import 'network.dart';

/// This is almost exact copy of https://github.com/anicdh/bitcoin_flutter/blob/master/lib/src/payments/p2pkh.dart
/// except using [Opcodes] static members instead of map
class P2PKH {
  P2PKHData data;
  Network network;

  P2PKH({@required data, network}) {
    this.network = network ?? Network.bitcoinCash();
    this.data = data;
    _init();
  }

  _init() {
    if (data.address != null) {
      _getDataFromAddress(data.address);
      _getDataFromHash();
    } else if (data.hash != null) {
      _getDataFromHash();
    } else if (data.output != null) {
      if (!isValidOutput(data.output))
        throw new ArgumentError('Output is invalid');
      data.hash = data.output.sublist(3, 23);
      _getDataFromHash();
    } else if (data.pubkey != null) {
      data.hash = hash160(data.pubkey);
      _getDataFromHash();
      _getDataFromChunk();
    } else if (data.input != null) {
      List<dynamic> _chunks = bscript.decompile(data.input);
      _getDataFromChunk(_chunks);
      if (_chunks.length != 2) throw new ArgumentError('Input is invalid');
      if (!bscript.isCanonicalScriptSignature(_chunks[0]))
        throw new ArgumentError('Input has invalid signature');
      if (!isPoint(_chunks[1]))
        throw new ArgumentError('Input has invalid pubkey');
    } else {
      throw new ArgumentError("Not enough data");
    }
  }

  void _getDataFromChunk([List<dynamic> _chunks]) {
    if (data.pubkey == null && _chunks != null) {
      data.pubkey = (_chunks[1] is int) ? new Uint8List.fromList([_chunks[1]]) : _chunks[1];
      data.hash = hash160(data.pubkey);
      _getDataFromHash();
    }
    if (data.signature == null && _chunks != null)
      data.signature = (_chunks[0] is int) ? new Uint8List.fromList([_chunks[0]]) : _chunks[0];
    if (data.input == null && data.pubkey != null && data.signature != null) {
      data.input = bscript.compile([data.signature, data.pubkey]);
    }
  }

  void _getDataFromHash() {
    if (data.address == null) {
      final payload = new Uint8List(21);
      payload.buffer.asByteData().setUint8(0, network.pubKeyHash);
      payload.setRange(1, payload.length, data.hash);
      data.address = bs58check.encode(payload);
    }
    if (data.output == null) {
      data.output = bscript.compile([
        Opcodes.OP_DUP,
        Opcodes.OP_HASH160,
        data.hash,
        Opcodes.OP_EQUALVERIFY,
        Opcodes.OP_CHECKSIG
      ]);
    }
  }

  void _getDataFromAddress(String address) {
    Uint8List payload = bs58check.decode(address);
    final version = payload.buffer.asByteData().getUint8(0);
    if (version != network.pubKeyHash)
      throw new ArgumentError('Invalid version or Network mismatch');
    data.hash = payload.sublist(1);
    if (data.hash.length != 20) throw new ArgumentError('Invalid address');
  }
}

class P2PKHData {
  String address;
  Uint8List hash;
  Uint8List output;
  Uint8List signature;
  Uint8List pubkey;
  Uint8List input;
  P2PKHData(
    {this.address,
      this.hash,
      this.output,
      this.pubkey,
      this.input,
      this.signature});

  @override
  String toString() {
    return 'P2PKHData{address: $address, hash: $hash, output: $output, signature: $signature, pubkey: $pubkey, input: $input}';
  }
}

isValidOutput(Uint8List data) {
  return data.length == 25 &&
    data[0] == Opcodes.OP_DUP &&
    data[1] == Opcodes.OP_HASH160 &&
    data[2] == 0x14 &&
    data[23] == Opcodes.OP_EQUALVERIFY &&
    data[24] == Opcodes.OP_CHECKSIG;
}


example/main.dart
import '../lib/babybitbox.dart' as Bitbox;

void main() async {
  // set this to false to use mainnet
  final testnet = true;

  // After running the code for the first time, depositing an amount to the address displayed in the console,
  // and waiting for confirmation, paste the generated mnemonic here,
  // so the code continues below with address withdrawal
  String mnemonic = "";

  if (mnemonic == "") {
    // generate 12-word (128bit) mnemonic
    mnemonic = Bitbox.Mnemonic.generate();

    print(mnemonic);
  }

  // generate a seed from mnemonic
  final seed = Bitbox.Mnemonic.toSeed(mnemonic);

  // create an instance of Bitbox.HDNode for mainnet
  final masterNode = Bitbox.HDNode.fromSeed(seed, testnet);

  // This format is compatible with Bitcoin.com wallet.
  // Other wallets use Change to m/44'/145'/0'/0
  final accountDerivationPath = "m/44'/0'/0'/0";

  // create an account node using the provided derivation path
  final accountNode = masterNode.derivePath(accountDerivationPath);

  // get account's extended private key
  final accountXPriv = accountNode.toXPriv();

  // create a Bitbox.HDNode instance of the first child in this account
  final childNode = accountNode.derive(0);

  // get an address of the child
  final address = childNode.toCashAddress();

  // if you are using testnet, set the appropriate rest api url before making
  // any API calls (like getting address or transaction details or broadcasting a transaction
  if (testnet) {
    Bitbox.Bitbox.setRestUrl(Bitbox.Bitbox.trestUrl);
  }

  // get address details
  final addressDetails = await Bitbox.Address.details(address);

  print(addressDetails);

  // If there is a confirmed balance, attempt to withdraw it to the address defined below
  if (addressDetails["balance"] > 0) {
    final builder = Bitbox.Bitbox.transactionBuilder(testnet: testnet);

    // retrieve address' utxos from the rest api
    final utxos = await Bitbox.Address.utxo(address) as List<Bitbox.Utxo>;

    // placeholder for input signatures
    final signatures = <Map>[];

    // placeholder for total input balance
    int totalBalance = 0;

    // iterate through the list of address utxos and use them as inputs for the withdrawal transaction
    utxos.forEach((Bitbox.Utxo utxo) {
      // add the utxo as an input for the transaction
      builder.addInput(utxo.txid, utxo.vout);

      // add a signature to the list to be used later
      signatures.add({
        "vin": signatures.length,
        "key_pair": childNode.keyPair,
        "original_amount": utxo.satoshis
      });

      totalBalance += utxo.satoshis;
    });

    // set an address to send the remaining balance to
    final outputAddress = "";

    // if there is an unspent balance, create a spending transaction
    if (totalBalance > 0 && outputAddress != "") {
      // calculate the fee based on number of inputs and one expected output
      final fee = Bitbox.BitcoinCash.getByteCount(signatures.length, 1);

      // calculate how much balance will be left over to spend after the fee
      final sendAmount = totalBalance - fee;

      // add the output based on the address provided in the testing data
      builder.addOutput(outputAddress, sendAmount);

      // sign all inputs
      signatures.forEach((signature) {
        builder.sign(signature["vin"], signature["key_pair"], signature["original_amount"]);
      });

      // build the transaction
      final tx = builder.build();

      // broadcast the transaction
      final txid = (await Bitbox.RawTransactions.sendRawTransaction([tx.toHex()])).first;

      // Yatta!
      print("Transaction broadcasted: $txid");
    } else if (totalBalance > 0) {
      print("Enter an output address to test withdrawal transaction");
    }
  }
}
